driver-core/src/main/com/mongodb/client/model/expressions/BranchesIntermediary.java

```diff
@@ -37,66 +43,197 @@ private static <T extends Expression> MqlExpression<?> mqlEx(final T value) {
+     * produces a value specified by the {@code mapping}.
+     *
+     * @param mapping the mapping.
+     * @return the appended sequence of checks.
+     */
+    public BranchesIntermediary<T, R> isNumber(final Function<? super NumberExpression, ? extends R> mapping) {
+        return is(v -> mqlEx(v).isNumber(), v -> mapping.apply((NumberExpression) v));
+    }
+
+    /**
+     * A successful check for
+     * {@linkplain Expression#isIntegerOr(IntegerExpression) being an integer}
+     * produces a value specified by the {@code mapping}.
+     *
+     * @param mapping the mapping.
+     * @return the appended sequence of checks.
+     */
+    public BranchesIntermediary<T, R> isInteger(final Function<? super IntegerExpression, ? extends R> mapping) {
+        return is(v -> mqlEx(v).isInteger(), v -> mapping.apply((IntegerExpression) v));
+    }
+
+    /**
+     * A successful check for
+     * {@linkplain Expression#isStringOr(StringExpression) being a string}
+     * produces a value specified by the {@code mapping}.
+     *
+     * @param mapping the mapping.
+     * @return the appended sequence of checks.
+     */
+    public BranchesIntermediary<T, R> isString(final Function<? super StringExpression, ? extends R> mapping) {
+        return is(v -> mqlEx(v).isString(), v -> mapping.apply((StringExpression) v));
+    }
+
+    /**
+     * A successful check for
+     * {@linkplain Expression#isDateOr(DateExpression) being a date}
+     * produces a value specified by the {@code mapping}.
+     *
+     * @param mapping the mapping.
+     * @return the appended sequence of checks.
+     */
+    public BranchesIntermediary<T, R> isDate(final Function<? super DateExpression, ? extends R> mapping) {
+        return is(v -> mqlEx(v).isDate(), v -> mapping.apply((DateExpression) v));
+    }
+
+    /**
+     * A successful check for
+     * {@linkplain Expression#isArrayOr(ArrayExpression) being an array}
+     * produces a value specified by the {@code mapping}.
+     *
+     * <p>Warning: The type argument of the array is not
+     * enforced by the API. The use of this method is an
+     * unchecked assertion that the type argument is correct.
+     *
+     * @param mapping the mapping.
+     * @return the appended sequence of checks.
+     * @param <Q> the type of the array.
+     */
     @SuppressWarnings("unchecked")
-    public <Q extends Expression> BranchesIntermediary<T, R> isArray(final Function<? super ArrayExpression<Q>, ? extends R> r) {
-        return is(v -> mqlEx(v).isArray(), v -> r.apply((ArrayExpression<Q>) v));
-    }
-
-    public BranchesIntermediary<T, R> isDocument(final Function<? super DocumentExpression, ? extends R> r) {
-        return is(v -> mqlEx(v).isDocumentOrMap(), v -> r.apply((DocumentExpression) v));
-    }
-
+    public <Q extends Expression> BranchesIntermediary<T, R> isArray(final Function<? super ArrayExpression<Q>, ? extends R> mapping) {
```

Comment:
The documentation of `Expression.isArrayOr` talks about "the type of elements of the ... array", while this documentation talks about the type argument of the parameterized type. Either approach works, but it would have been better if the documentation were consistent.